### PR TITLE
[BACKLOG-17884] Make ThinConnection.localClient binding independent of Spoon's lifecycle

### PR DESCRIPTION
**Pull request by nantunes**
@pentaho/millenniumfalcon please review.

### CODE REVIEW

**Comment by nantunes** [comment]
@pentaho/bb-8 please review.

File: pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/DataServiceLocalClientBinder.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  public void bind( DataServiceClientService service ) {
+    dataServiceClientService.set( service );
+
+    ThinConnection.localClient = service;
+  }
+
+  public void unbind( DataServiceClientService service ) {
+    if ( dataServiceClientService.compareAndSet( service, null ) ) {
+      dataServiceClientService.set( null );
```
Comment (graimundo) [inline pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/DataServiceLocalClientBinder.java]: @nantunes doesn't the `compareAndSet` already do what line 41 is doing?

Comment (nantunes) [reply pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/DataServiceLocalClientBinder.java]: Well spot. :-)

File: pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/SpoonDataServiceLocalClientConfigurer.java [hunk 2]
```diff
@@ -81,9 +85,7 @@ private synchronized void setup( DataServiceClientService clientService ) {
 
       clientService.setRepository( spoon.getRepository() );
       clientService.setMetaStore( spoon.getMetaStore() );
-      ThinConnection.localClient = clientService;
-    } else {
-      ThinConnection.localClient = null;
     }
+    // else, should we clean (set to null) the clientService repository and metastore?
```
Comment (graimundo) [inline pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/SpoonDataServiceLocalClientConfigurer.java:89]: @mkambol any idea if we should set the repository and metastore to null?

**Review by graimundo — APPROVED** [review]
Looks good.

File: pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/SpoonDataServiceLocalClientConfigurer.java (hunk 2 above)
Comment (mkambol) [reply pdi-dataservice-client/src/main/java/org/pentaho/di/trans/dataservice/client/SpoonDataServiceLocalClientConfigurer.java:89]: Scanning through the code it doesn't appear those values are even used any more.  Both the .getRepository and .getMetastore methods are marked deprecated in RemoteClient and appear to be no-ops in the implementations.  I think "locator" classes are used to retrieve metastore and repos now.

So yeah, you might be able to clean this out more and eliminate some unnecessary code.  If it looks too risky or uncertain you could leave as is and add a comment to revisit.

**Pull request merged**